lib/mox/server.ex

```diff
@@ -67,20 +73,28 @@ defmodule Mox.Server do
     {:reply, pending, state}
   end
 
+  def handle_call({:allow, mock, owner_pid, pid}, _from, state) do
+    state =
+      put_in(state[:allowances][{mock, pid}], owner_pid)
+
+    {:reply, {:ok, pid}, state}
+  end
+
   def handle_info({:DOWN, _, _, pid, _}, state) do
-    {:noreply, Map.delete(state, pid)}
+    {_, state} = pop_in(state[:expectations][pid])
+    {:noreply, state}
   end
 
   # Helper functions
 
   defp maybe_add_and_monitor_pid(state, pid) do
     case state do
```

Comment:
Cosmetic but I would do `state.expectations` and continue matching only on the pid.